util/test/sub_test.py

```diff
@@ -740,1716 +741,1756 @@ def filter_errors(output_in, pre_exec_output, execgoodfile, execlog):
+    if not os.access(compiler,os.R_OK|os.X_OK):
+        Fatal('Cannot execute compiler \''+compiler+'\'')
+
+    path_to_compiler=os.path.abspath(os.path.dirname(compiler))
+    # Assume chpl binary is 2 directory levels down in the base installation
+    (chpl_base, tmp) = os.path.split(path_to_compiler)
+    (chpl_base, tmp) = os.path.split(chpl_base)
+    chpl_base=os.path.normpath(chpl_base)
+    # sys.stdout.write('CHPL_BASE='+chpl_base+'\n')
+    return chpl_base
+
+def get_chpl_home(chpl_base):
+    """Returns the normalized path to the Chapel installation."""
+    # If $CHPL_HOME is not set, use the base installation of the compiler
+    chpl_home=os.getenv('CHPL_HOME', chpl_base);
+    chpl_home=os.path.normpath(chpl_home)
+    # sys.stdout.write('CHPL_HOME='+chpl_home+'\n');
+    return chpl_home
+
+def main(): 
+    # Start of sub_test proper
+    #
+    global utildir, chpl_base, chpl_home, machine, envCompopts, platform
+    global chplcommstr, chplnastr, chpllmstr, chpltasksstr
+
+    if len(sys.argv)!=2:
+        print('usage: sub_test COMPILER')
+        sys.exit(0)
+
+    compiler=sys.argv[1]
+    is_chpldoc = compiler.endswith('chpldoc')
+    chpl_base = get_chpl_base(compiler)
+    chpl_home = get_chpl_home(chpl_base)
+    localdir = get_local_dir(chpl_base)
+
+    # Find the test util directory -- set this in start_test to permit
+    # a version of start_test other than the one in CHPL_HOME to be used
+    utildir=os.getenv('CHPL_TEST_UTIL_DIR');
+    if utildir is None or not os.path.isdir(utildir):
+        Fatal('Cannot find test util directory {0}'.format(utildir))
+
+    # Needed for MacOS mount points
+    utildir = os.path.realpath(utildir)
+    # sys.stdout.write('utildir='+utildir+'\n');
+
+    # Find the c compiler
+    # We open the compileline inside of CHPL_HOME rather than CHPL_TEST_UTIL_DIR on
+    # purpose. compileline will not work correctly in some configurations when run
+    # outside of its directory tree.
+    compileline = os.path.join(chpl_home, 'util', 'config', 'compileline')
+    def run_compileline(flag, lookingfor):
+        (returncode, result, _) = run_process([compileline, flag],
+                                              stdout=subprocess.PIPE,
+                                              stderr=subprocess.PIPE)
+        result = result.rstrip()
+        if returncode != 0:
+          Fatal('Cannot find ' + lookingfor)
+        return result
+
+    c_compiler = run_compileline('--compile', 'c compiler')
+    cpp_compiler = run_compileline('--compile-c++', 'c++ compiler')
+    host_c_compiler = run_compileline('--host-c-compiler', 'host c compiler')
+    host_cpp_compiler = run_compileline('--host-cxx-compiler', 'host c++ compiler')
+    runtime_includes_and_defines = run_compileline('--includes-and-defines',
+                                                   'runtime includes and defines')
+
+    # Find the test directory
+    testdir=chpl_home+'/test'
```

Comment:
Is this code (and subsequent lines) duplicate now? It wasn't before -- could the `test_root_dir` computation be factored out?

Reply:
Good suggestion, I've done some additional refactoring.